odoo/modules/registry.py

```diff
@@ -467,15 +467,30 @@ def check_indexes(self, cr, model_names):
         if not expected:
             return
 
+        trgm = sql.has_pg_trgm(cr)
         cr.execute("SELECT indexname FROM pg_indexes WHERE indexname IN %s",
                    [tuple(row[0] for row in expected)])
         existing = {row[0] for row in cr.fetchall()}
 
-        for indexname, tablename, columnname, index in expected:
-            if index and indexname not in existing:
+        for indexname, model, columnname, field in expected:
+            tablename = model._table
+            index = field.index
+            assert index in ('btree', 'gin', 'not null', True, False)
+            if field.index and indexname not in existing:
+                where = ''
+                method = 'btree'
+                operator = ''
+                if index == 'not null':
+                    where = ' WHERE "%s" IS NOT NULL' % columnname
+                if index == 'gin':
+                    if trgm:
+                        operator = 'gin_trgm_ops'
+                        method = 'gin'
+                    else:
+                        method = "btree"
```

Comment:
Having reached this point of specificity, why not make `index` similar to `order` and allow the user to indicate wathever condition (s)he considers more appropriate in SQL?, e.g.: 

```python
payment_id = fields.Many2one(
    index='payment_id IS NOT NULL',
```

but also:

```python
payment_id = fields.Many2one(
    index='payment_id IS NOT NULL AND is_reconciled IS TRUE',
```

Thus would be:

```suggestion
                if index == 'gin':
                    if trgm:
                        operator = 'gin_trgm_ops'
                        method = 'gin'
                    else:
                        method = "btree"
                elif index:
                    where = ' WHERE ' % index
```

Reply:
> Having reached this point of specificity, why not make `index` similar to `order` and allow the user to indicate wathever condition (s)he considers more appropriate in SQL?

NOT NULL is really the main / only use case because it's directly related to an internal mechanism of Odoo: stats button, or one2many fields. They are used in queries related to their relation, so when it's not null. (e.g. stats on a partner or visitor). It's the "elephant in the room" compared to more specific use cases.

More specific use cases are far from being obvious: for example, in your example, people may filter on entries that are not reconciled, so the index would not be effiicent.